### PR TITLE
Refactor ReferenceList

### PR DESCRIPTION
**Pull request by Benestar**
This removes the inheritance between ReferenceList and SplObjectStorage
and replaces it with a native array. To make the change as minimal as
possible, the Countable interface and the IteratorAggregate interface are
now implemented by ReferenceList so that it can be used almost the same as
before (like iterating over it and calling count() on it).

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
Needs rebase

**Comment by JeroenDeDauw** [comment]
Includes #608

**Comment by JeroenDeDauw** [comment]
Rebased

File: src/ReferenceList.php [hunk 1]
```diff
@@ -104,26 +99,7 @@ public function addNewReference( $snaks = array() /*...*/ ) {
...
-			$this->detach( $object );
-		}
-
-		// Attach the new reference and reattach the previously detached references:
-		$this->attach( $reference );
-
-		foreach ( $referencesToShift as $object ) {
-			$this->attach( $object );
-		}
+		array_splice( $this->references, $index, 0, array( $reference ) );
```
Comment (JeroenDeDauw) [inline src/ReferenceList.php]: Hah, this is a lot nicer!

File: src/ReferenceList.php [hunk 2]
```diff
@@ -286,4 +263,35 @@ public function equals( $target ) {
...
 	}
 
+	/**
+	 * @see Countable::count
+	 *
+	 * @since 5.0
+	 *
+	 * @return int
+	 */
+	public function count() {
```
Comment (JeroenDeDauw) [inline src/ReferenceList.php]: This one is not new

**Comment by JeroenDeDauw** [comment]
But but but! `SplObjectStorage` is more efficient!

+2, when follow up is merged

**Comment by Benestar** [comment]
> But but but! SplObjectStorage is more efficient!

Will make a huge difference for our immense lists of references ;p

File: src/ReferenceList.php [hunk 3]
```diff
@@ -150,13 +125,10 @@ public function hasReference( Reference $reference ) {
 	 * @return int|boolean
 	 */
 	public function indexOf( Reference $reference ) {
-		$index = 0;
-
-		foreach ( $this as $object ) {
-			if ( $object === $reference ) {
+		foreach ( $this->references as $index => $ref ) {
+			if ( $ref->equals( $reference ) ) {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Is this breaking change (from identity to equality) documented?

File: src/ReferenceList.php [hunk 4]
```diff
@@ -59,31 +62,23 @@ public function __construct( $references = array() ) {
 	 * @throws InvalidArgumentException
 	 */
 	public function addReference( Reference $reference, $index = null ) {
-		if ( !is_int( $index ) && $index !== null ) {
-			throw new InvalidArgumentException( '$index must be an integer or null' );
+		if ( $index !== null && ( !is_int( $index ) || $index < 0 ) ) {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: An other breaking change that must be documented. There should also be a test for this.

File: src/ReferenceList.php [hunk 5]
```diff
@@ -59,31 +62,23 @@ public function __construct( $references = array() ) {
...
 		}
 	}
 
 	/**
-	 * @see SplObjectStorage::attach
-	 *
-	 * @param Reference $reference
-	 * @param mixed $data Unused in the ReferenceList class.
-	 */
-	public function attach( $reference, $data = null ) {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: In my opinion all removed methods (this also includes `contains`, `getHash` and more) should be listed in the release notes.

File: src/ReferenceList.php [hunk 6]
```diff
@@ -50,6 +52,7 @@ public function __construct( $references = array() ) {
 
 	/**
 	 * Adds the provided reference to the list.
+	 * Empty references are ignored.
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Missing in the release notes.

File: src/ReferenceList.php [hunk 7]
```diff
@@ -286,4 +263,33 @@ public function equals( $target ) {
...
+	}
+
+	/**
+	 * @see IteratorAggregate::getIterator
+	 *
+	 * @since 5.0
+	 *
+	 * @return Traversable
+	 */
+	public function getIterator() {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Release notes? Test? Why was this added, in a "refactoring" patch?

File: src/ReferenceList.php [hunk 8]
```diff
@@ -286,4 +263,33 @@ public function equals( $target ) {
...
+		return new ArrayIterator( $this->references );
+	}
+
+	/**
+	 * @since 5.0
+	 *
+	 * @return Reference[]
+	 */
+	public function toArray() {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Release notes? Test? Why was this added, in a "refactoring" patch?

File: src/ReferenceList.php [hunk 9]
```diff
@@ -286,4 +263,33 @@ public function equals( $target ) {
...
+	 * @return Traversable
+	 */
+	public function getIterator() {
+		return new ArrayIterator( $this->references );
+	}
+
+	/**
+	 * @since 5.0
+	 *
+	 * @return Reference[]
```
Comment (thiemowmde) [inline src/ReferenceList.php]: "Numerically indexed (non-sparse) array."

File: src/ReferenceList.php [hunk 10]
```diff
@@ -194,15 +166,23 @@ public function hasReferenceHash( $referenceHash ) {
...
+				$this->removeReferenceAtIndex( $index );
+			}
 		}
 	}
 
 	/**
-	 * Returns the reference with the provided hash, or null if there is no such reference in the list.
+	 * @param int $index
+	 */
+	private function removeReferenceAtIndex( $index ) {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Typically I find single-line methods like this one confusing and not helpful. Did you planned to make this public?

File: src/ReferenceList.php [hunk 11]
```diff
@@ -211,12 +191,9 @@ public function removeReferenceHash( $referenceHash ) {
...
 	 */
 	public function getReference( $referenceHash ) {
-		/**
-		 * @var Hashable $hashable
-		 */
-		foreach ( $this as $hashable ) {
-			if ( $hashable->getHash() === $referenceHash ) {
-				return $hashable;
+		foreach ( $this->references as $reference ) {
+			if ( $reference->getHash() === $referenceHash ) {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: There are two loops in this class that search for a hash, but no method to access elements by index. Therefore it would probably be more efficient (less code, maybe faster) to index the array by hash, and not numerically.

File: tests/unit/ReferenceListTest.php [hunk 12]
```diff
@@ -372,14 +372,4 @@ public function testGivenNonEmptyList_isNotEmpty() {
 		$this->assertFalse( $references->isEmpty() );
 	}
 
-	public function testGivenNonEmptyListWithForwardedIterator_isNotEmpty() {
-		$references = new ReferenceList();
-		$references->addNewReference( new PropertyNoValueSnak( 1 ) );
-		$references->next();
```
Comment (thiemowmde) [inline tests/unit/ReferenceListTest.php]: `next` is not removed, as far as I can see. Why is this test removed then?

Comment (Benestar) [inline tests/unit/ReferenceListTest.php]: Afaik `next` got removed? /me is confused

File: src/ReferenceList.php (hunk 5 above)
Comment (Benestar) [inline src/ReferenceList.php]: I don't think it makes sense to list all the methods `SplObjectStorage` contains (http://php.net/manual/en/class.splobjectstorage.php). Isn't it enough to note that `ReferenceList` no longer extends that class?

File: src/ReferenceList.php (hunk 11 above)
Comment (Benestar) [inline src/ReferenceList.php]: Good point, we can do this in another patch.

File: src/ReferenceList.php (hunk 7 above)
Comment (Benestar) [inline src/ReferenceList.php]: Intention is that we don't break loops that iterate over a `ReferenceList`. `foreach ( $referenceList as $reference ) {` should still work with this.

File: src/ReferenceList.php (hunk 4 above)
Comment (Benestar) [inline src/ReferenceList.php]: Done

File: src/ReferenceList.php (hunk 3 above)
Comment (Benestar) [inline src/ReferenceList.php]: Fixed

File: tests/unit/ReferenceListTest.php (hunk 12 above)
Comment (thiemowmde) [inline tests/unit/ReferenceListTest.php]: `next` is in the `Iterator` interface, which is not removed, because `ReferenceList` still `implements IteratorAggregate`.

Comment (Benestar) [inline tests/unit/ReferenceListTest.php]: `IteratorAggregate` does not extend the `Iterator` interface but only `Traversable`. (`IteratorAggregate extends Traversable`) I tested it manually and I cannot call a `next` method on `ReferenceList. (`Call to undefined method Wikibase\DataModel\ReferenceList::next()`)

File: src/ReferenceList.php (hunk 5 above)
Comment (thiemowmde) [inline src/ReferenceList.php]: That's exactly what I said, why I feel so lost and what I consider the top one blocking issue in releasing this: Without detailed release notes it's extremely hard, if not impossible to update code that uses this class.

Also see my comments in #603.

Comment (Benestar) [inline src/ReferenceList.php]: So you propose to add the complete list of functions in `SplObjectStorage` to the release notes? That might indeed help to see which methods exactly got removed from the public interface of `ReferenceList`.

Your latest comment? I didn't tell Jeroen to merge that patch. Should I add a "DNM until three persons have seen this" or "DNM for 3 hours" in the description of my pull requests?

File: tests/unit/ReferenceListTest.php (hunk 12 above)
Comment (thiemowmde) [inline tests/unit/ReferenceListTest.php]: Oh, wow. Is this an other line missing in the release notes? "ReferenceList does not implement Iterator any more". Including "ReferenceList does not implement current, next, key, valid and rewind any more" to make it absolutely clear what this means.

Comment (Benestar) [inline tests/unit/ReferenceListTest.php]: See https://github.com/wmde/WikibaseDataModel/pull/609/files#r52314875

In release notes we say "ReferenceList no longer derives from SplObjectStorage, though still implements Countable and Traversable" which implies it does no longer implement the other interfaces of SplObjectStorage. Also please avoid comments like "Oh, wow" which are kind of disrespectful. We are not entirely stupid...

File: src/ReferenceList.php (hunk 5 above)
Comment (thiemowmde) [inline src/ReferenceList.php]: > add the complete list of functions in SplObjectStorage to the release notes?

No. Why should one do that? List methods that can not be called any more.

File: tests/unit/ReferenceListTest.php (hunk 12 above)
Comment (thiemowmde) [inline tests/unit/ReferenceListTest.php]: Code typically does not type hint against SplObjectStorage or Countable or Traversable. Sure, all this should be mentioned in the release notes, but is barely helpful.

"Oh, wow" was meant as "oh, wow, PHP is broken". I reviewed the changes that touch this code a long, long time ago and I'm looking at them again for almost a full day now, and I still don't fully understand what's going on. There is a reason I never merged these patches. Verbose, crystal clear documentation is absolutely crucial here.
